### PR TITLE
add ODM parsing for study gender and dateOfBirth requirements

### PR DESCRIPTION
**Pull request by beaugogh**

### CODE REVIEW

**Comment by PiotrZakrzewski** [comment]
Can you remove jar from the commit?
EDIT: a, wait no that is fine

File: src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+            List<ClinicalData> clinicalDatas = clinicalDataRepository.findBySubmission(uploadSession);
+
+            Map<String, String> subjectMap = openClinicaService.createMapSubjectLabelToSubjectOID(username, pwdHash, url, clinicalDatas);
+            List<String> test = new ArrayList<>();
+            for(String key: subjectMap.keySet()) {
+                test.add(key);
+            }
+            return new ResponseEntity<>(test, HttpStatus.OK);
+
+//            return new ResponseEntity<>(HttpStatus.OK);
```
Comment (PiotrZakrzewski) [inline src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java]: This line can be removed

File: src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    public ResponseEntity<List<String>> checkNewPatients(HttpSession session) {
+        try {
+            UploadSession uploadSession = uploadSessionService.getCurrentUploadSession(session);
+            OcUser user = ocUserService.getCurrentOcUser(session);
+            String username = user.getUsername();
+            String pwdHash = ocUserService.getOcwsHash(session);
+            String url = user.getOcEnvironment();
+            List<ClinicalData> clinicalDatas = clinicalDataRepository.findBySubmission(uploadSession);
+
+            Map<String, String> subjectMap = openClinicaService.createMapSubjectLabelToSubjectOID(username, pwdHash, url, clinicalDatas);
```
Comment (PiotrZakrzewski) [inline src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java:56]: Does the SOAP call behind it work?

File: src/main/java/nl/thehyve/ocdu/models/OcDefinitions/SiteDefinition.java [hunk 3]
```diff
@@ -20,7 +20,12 @@
 
     private boolean isGenderRequired = true;
 
-    private boolean isBirthdateRequired = true;
+    /*
+     * 1. yes, required
+     * 2. only year of birth
```
Comment (PiotrZakrzewski) [inline src/main/java/nl/thehyve/ocdu/models/OcDefinitions/SiteDefinition.java:25]: Are you sure it is allowed to register patient with full date (more than only year)|?

File: src/main/java/nl/thehyve/ocdu/soap/ResponseHandlers/GetStudyMetadataResponseHandler.java [hunk 4]
```diff
@@ -138,6 +141,102 @@ public static MetaData parseGetStudyMetadataResponse(SOAPMessage response) throw
...
+            }//for
+        }//if
+
+        return isGenderRequired;
+    }
+
+    private static int parseStudyDateOfBirthRequired(Document odm, String mypath) throws XPathExpressionException {
+        /*
+         * 1. yes, required
+         * 2. only year of birth
```
Comment (PiotrZakrzewski) [inline src/main/java/nl/thehyve/ocdu/soap/ResponseHandlers/GetStudyMetadataResponseHandler.java:205]: Was this gradation from ODM or from our Design document?

**Comment by PiotrZakrzewski** [comment]
No new tests and yet test coverage increases ;) 
I guess this is because metadata class is extended here and it is covered with tests already.

File: src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java (hunk 1 above)
Comment (beaugogh) [inline src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java]: yes, removed:)

File: src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java (hunk 2 above)
Comment (beaugogh) [inline src/main/java/nl/thehyve/ocdu/controllers/TemplateController.java:56]: nope, the SOAP call createMapSubjectLabelToSubjectOID  still needs to be revised with Jacob on Monday, I added a TODO comment as a reminder.

File: src/main/java/nl/thehyve/ocdu/models/OcDefinitions/SiteDefinition.java (hunk 3 above)
Comment (beaugogh) [inline src/main/java/nl/thehyve/ocdu/models/OcDefinitions/SiteDefinition.java:25]: yes, actually there is an integer [1,3] indicating 1.full date, 2.only year, and 3.not used, I have changed the code correspondingly, please check.

**Comment by beaugogh** [comment]
yeah, I was about to ask that question. Weird, but good:)

File: src/main/java/nl/thehyve/ocdu/soap/ResponseHandlers/GetStudyMetadataResponseHandler.java (hunk 4 above)
Comment (beaugogh) [inline src/main/java/nl/thehyve/ocdu/soap/ResponseHandlers/GetStudyMetadataResponseHandler.java:205]: you mean /*
-         \* 1. yes, required
-         \* 2. only year of birth ... ?

this comment is sth. I summarized from the sjorgen odm file, which was not in the design doc.

**Comment by beaugogh** [comment]
hmm, some tests are failing, no sure what goes wrong. Local build is fine...

**Comment by PiotrZakrzewski** [comment]
Tests are passing - it is just travis node running out of memory, after rerunning it will pass. I will try to fix it today.

**Comment by PiotrZakrzewski** [comment]
Okay fixed. Merge #111  and rebase your branch to get this patch. Now travis seems to behave better.

**Comment by beaugogh** [comment]
merged #111

File: src/main/java/nl/thehyve/ocdu/validators/patientDataChecks/DateOfBirthPatientDataCheck.java [hunk 5]
```diff
@@ -19,38 +20,72 @@
     @Override
     public ValidationErrorMessage getCorrespondingError(int index, Subject subject, MetaData metaData) {
 
+        int DOBrequired = metaData.getBirthdateRequired();
+        for (int i = 0; i < metaData.getSiteDefinitions().size(); i++) {
+            SiteDefinition sd = metaData.getSiteDefinitions().get(i);
+            int site_dob = sd.getBirthdateRequired();
+            //if site requirement for dateOfBirth is more specific than that of study,
```
Comment (PiotrZakrzewski) [inline src/main/java/nl/thehyve/ocdu/validators/patientDataChecks/DateOfBirthPatientDataCheck.java:27]: great stuff!

**Pull request merged**